### ISSUE TITLE
Manufacturer error when upgrading to 2.2.6

### ISSUE THREAD

**Issue by N1ghtfl0w** [magento/magento2#18134]
### Preconditions
1. Magento 2.2.5

### Steps to reproduce
1. Upgrade to Magento 2.2.6

### Expected result
1. Upgrade succesfull

### Actual result
1. Getting error from missing attribute Manufacturer

----------------------------------------------------------------------

I updated to Magento 2.2.6 via composer this morning however when running bin/magneto setup:upgrade I get the following error:

Attribute with ID "Manufacturer" does not exist

This shows after the Upgrading data part for Magento_ConfigurableProduct within the setup upgrade command. Am wondering how to fix this and or if anyone else is experiencing it?

This seems to be caused by below code within UpgradeData.php for the Configurable Product module:

```
   if (version_compare($context->getVersion(), '2.2.1') < 0) {
        $relatedProductTypes = $this->getRelatedProductTypes('manufacturer', $eavSetup);
        if (!in_array(Configurable::TYPE_CODE, $relatedProductTypes)) {
            $relatedProductTypes[] = Configurable::TYPE_CODE;
            $this->updateRelatedProductTypes('manufacturer', $relatedProductTypes, $eavSetup);
        }
    }
```
I do not have a manufacturer attribute within this instance and I'm wondering if it has been deleted. Looking at other instances however this attribute is present but not a system attribute and can be deleted from admin so am wondering if this attribute is really needed and if not this should not really cause setup:upgrade to break?

**Comment by magento-engcom-team** [comment]
Hi @N1ghtfl0w. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@N1ghtfl0w do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `acknowledged` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by ghost** [comment]
Hi @N1ghtfl0w thank you for you report,  I'm not able to reproduce following steps you described, Can you provide more information about have you some additional data or some else ?.
![selection_115](https://user-images.githubusercontent.com/40756439/45760267-59749780-bc32-11e8-9fd4-73d5cdc47209.png)

**Comment by hostep** [comment]
This is very easily reproducible, I understand that the steps to reproduce aren't 100% accurate, but just reading @N1ghtfl0w's post it's easy to see what goes wrong.

So here are the exact steps to reproduce:
1. Setup a clean Magento 2.2.5 installation (I used composer)
2. Open the backend
3. Go to Stores > Attributes > Products
4. Search the 'manufacturer' attribute and delete it
5. Open your composer.json file and change the version constraint of the `magento/product-community-edition` package from `2.2.5` to `2.2.6`
6. Run `composer update`
7. Run `bin/magento setup:upgrade`

Result:
```
...
Module 'Magento_Quote':
Upgrading data..
Module 'Magento_SalesSequence':
Module 'Magento_Security':
Module 'Magento_CatalogImportExport':
Module 'Magento_Sales':
Upgrading data..
Module 'Magento_Cron':
Module 'Magento_Msrp':
Module 'Magento_Search':
Module 'Magento_CatalogUrlRewrite':
Module 'Magento_Widget':
Module 'Magento_CatalogInventory':
Module 'Magento_Checkout':
Module 'Magento_Vault':
Module 'Magento_CmsUrlRewrite':
Module 'Magento_User':
Module 'Magento_ConfigurableImportExport':
Module 'Magento_ConfigurableProduct':
Upgrading data.. Attribute with ID: "manufacturer" does not exist
```

**Comment by N1ghtfl0w** [comment]
@hostep thank you for the exact steps to reproduce! So if im correct, if we add the manufacturer attribute again in 2.2.5 the update will pass? 

And if so, what are the correct settings for this attribute?

**Comment by hostep** [comment]
Looks like it indeed.
For the correct settings, I don't know exactly, but if you setup a new clean 2.2.5 installation, you can probably distil it from there.

**Comment by ghost** [comment]
Hi @N1ghtfl0w the fix for this issue avaible in 2.3-develop branch, Hi @hostep Thank you for explanations.

**Comment by hostep** [comment]
@engcom-backlog-nazar: this NEEDS to be fixed in 2.2, you just can't ignore this, this is a database migration script we are talking about here.

The fact that it works in 2.3-develop is probably because the database migration scripts are completely different there.

Re-opening.

**Comment by ghost** [comment]
@hostep ok.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95113` were created

**Comment by ghost** [comment]
@N1ghtfl0w, thank you for your report.
We've acknowledge the issue and added to our backlog.

**Comment by N1ghtfl0w** [comment]
For the ones that are searching for a quick fix:

Just add the manufacturer attribute under "Stores > Attributes > Product > Add New Attribute" as a dropdown attribute and set the scope, under  Advanced Attribute Properties, to global.

After this we did the update and it was succesfull.

**Comment by harrigo** [comment]
This happened to me also. We do not use the Manufacturer attribute and I can add it back to do the update however am wondering if I can delete it again after this update or is it required and should it really be a system attribute that cannot be deleted if that is the case?

**Comment by MartinAarts** [comment]
We have a related problem. Attribute does still exist after upgrading from 2.2.5 to 2.2.6 but doesn't show in backend. Shows as usual in the frontend.

**Comment by ghost** [comment]
> 
> 
> We have a related problem. Attribute does still exist after upgrading from 2.2.5 to 2.2.6 but doesn't show in backend. Shows as usual in the frontend.

I am experiencing the same issue. The update from 2.2.5 to 2.2.6 completed successfully but the manufacturer attribute does not appear when editing a product on the backend. The manufacturer attribute shows on the frontend product page for products that were created before the 2.2.6 upgrade but does not appear for products created after. It no longer shows in the backend for any products; it still is listed under the Stores > Attributes > Product page.

**Comment by happydownload33** [comment]
bug is created by Magento themself i you can see it on release log 2.2.6
Magento now displays the manufacturer attribute on the Admin on the Catalog page for configurable products. Previously, Magento displayed these attributes on the simple products catalog page, but not on the configurable products catalog page.

i think that they forget to test the Simple product module after editing the files

well if you do add product in backend and then configurable product manufacturer is showing in the list.
but when you fill the fields and do add product it disappear.

**Comment by bengower** [comment]
Anyone know the commit ref for this change that has led to the manufacturer attribute disappearing from simple products?

**Comment by ajusev** [comment]
Seems that for whatever reason during the upgrade from 2.2.5 to 2.2.6, new code did not actually preserved list of existing product types when updating the manufacturers plugin.

Looking at the catalog_eav_attribute table for the attribute_id that equals to your manufacturer attribute (you can look this up using: 
`select attribute_id from eav_attribute where attribute_code = 'manufacturer';`
) - The value of the **apply_to** column in the catalog_eav_attribute has ",configurable" which only causes to be shown in configurable products.

You can use following to apply quick work around for it:
`update catalog_eav_attribute set apply_to = 'simple,virtual,bundle,downloadable,configurable' where attribute_id = YOUR_MANUFACTURER_ATTR_ID;`

**Comment by MartinAarts** [comment]
> Seems that for whatever reason during the upgrade from 2.2.5 to 2.2.6, new code did not actually preserved list of existing product types when updating the manufacturers plugin.
> 
> Looking at the catalog_eav_attribute table for the attribute_id that equals to your manufacturer attribute (you can look this up using:
> `select attribute_id from eav_attribute where attribute_code = 'manufacturer';`
> ) - The value of the **apply_to** column in the catalog_eav_attribute has ",configurable" which only causes to be shown in configurable products.
> 
> You can use following to apply quick work around for it:
> `update catalog_eav_attribute set apply_to = 'simple,virtual,bundle,downloadable,configurable' where attribute_id = YOUR_MANUFACTURER_ATTR_ID;`

Great, this fixed it for us!

**Comment by JulianZetch** [comment]
@ajusev 

> 
> 
> Seems that for whatever reason during the upgrade from 2.2.5 to 2.2.6, new code did not actually preserved list of existing product types when updating the manufacturers plugin.
> 
> Looking at the catalog_eav_attribute table for the attribute_id that equals to your manufacturer attribute (you can look this up using:
> `select attribute_id from eav_attribute where attribute_code = 'manufacturer';`
> ) - The value of the **apply_to** column in the catalog_eav_attribute has ",configurable" which only causes to be shown in configurable products.
> 
> You can use following to apply quick work around for it:
> `update catalog_eav_attribute set apply_to = 'simple,virtual,bundle,downloadable,configurable' where attribute_id = YOUR_MANUFACTURER_ATTR_ID;`

We had the same issue. We have upgraded our Magento 2 version to 2.2.6 yesterday and the manufacturer attribute wasn't able to create new products. Fortunately, with the update that you shared us, it was resolved. We can create new products and assign them their manufacturer, but if we want to change the manufacturer after creating the product, the manufacturer attribute isn't able anymore.

Anyway, it was so helpful. Best regards.

**Comment by utklasad** [comment]
> Seems that for whatever reason during the upgrade from 2.2.5 to 2.2.6, new code did not actually preserved list of existing product types when updating the manufacturers plugin.
> 
> Looking at the catalog_eav_attribute table for the attribute_id that equals to your manufacturer attribute (you can look this up using:
> `select attribute_id from eav_attribute where attribute_code = 'manufacturer';`
> ) - The value of the **apply_to** column in the catalog_eav_attribute has ",configurable" which only causes to be shown in configurable products.
> 
> You can use following to apply quick work around for it:
> `update catalog_eav_attribute set apply_to = 'simple,virtual,bundle,downloadable,configurable' where attribute_id = YOUR_MANUFACTURER_ATTR_ID;`

This fixed my issue. The manufacturer attribut disappeared on all new products after upgrade to 2.2.6 from 2.2.5. All old products had the manufacturer attribut and I couldn't set it on new ones.

Thanks buddy!

**Comment by utklasad** [comment]
FYI -

It's probably this commit: https://github.com/magento/magento2/commit/1ddc94b84f357cb56916baaba326ccf65c8372a4

Thats removing the simple product type from the apply_to column to the manufacturer attribute in the database. I reverted this file to the 2.2.5 version and the issue disappeared.

**Comment by gewaechshaus** [comment]
Confirmed, trying to fix.

**Comment by nickpiro** [comment]
Anyone fix this in 2.2.6? I'm seeing a lot of errors in 2.2.6. Should I update to 2.2.7? I'm not sure, but a lot of people use manufacturer attribute!

**Comment by hostep** [comment]
@nickpiro: this issue hasn't been fixed in 2.2.7, but there is a pending PR which should fix it over here: https://github.com/magento/magento2/pull/19551.

Btw: this issue is not for people who **use** the manufacturer attribute, but for people who **removed** it, if you still have the manufacturer attribute, then everything should be fine :)

**Comment by nickpiro** [comment]
@hostep we use the manufacturer attribute, we never removed it. Updated from 2.2.5 to 2.2.6 and now the attribute is not visible on the product edit page in the backend. The manufacturer attribute is still there, the manufacturers are still being associated to the original products that had them associated, but when you go into edit or add a product, the manufacturer attribute is no longer visible in the product edit page. When you look at the Attribute Set page, and select an attribute set, you see manufacturer attribute is still listed under the 'Attributes' folder, but does not display on the product edit page.

**Comment by hostep** [comment]
@nickpiro: aha, I wasn't aware of this other bug. It might be best if you open a new issue then. This issue here is for when people remove the manufacturer attribute, that the database migrations crash, because they expect the attribute to exist.

**Comment by nickpiro** [comment]
@hostep thanks man! Haha. I will do this now. 👍

**Comment by justinpfister** [comment]
We saw this issue on 2.2.6 where we could not edit manufacturer_id in the backend AND we used manufacturer id.. any talk of deleting the attribute didn't sound like a great idea. We went with the upgrade to 2.2.7 and it immediately fixed our issue. We can now edit manufacturer on the product pages. Thank you everyone on this thread.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-tomash. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by ghost** [comment]
Closing per comments and because of #19514

**Comment by ghost** [comment]
Strange tom see that from a clean magento 2.3 installation i still have common issues like this.
In my case I was advised by a message in magento admin to use Elastic search in stead of the default search engine of magento. If I look from the non programmer perspective it is running into the first problem following this advice.

In the end there is no way a Non programmer can find out what is going on here.
And don't mention how to reproduce such problems.
So if you want to see what is happening even if you don't understand it.
You can:

copy public_html/pub/errors/local.xml.sample
To  public_html/pub/errors/local.xml

And post this result on the forums without a server name.
I changed my server name to **myserver**.

1. If i simply click on a menu item (category) this message is popping up.

Notice: Undefined index: **manufacturer**_bucket in /home/**myserver**/public_html/vendor/magento/module-elasticsearch/SearchAdapter/Aggregation/Builder/Term.php on line 23

2 the manufacturer attribute is there by default in magento 2.3 as a drop-down.

I can not confirm but I can assume everyone with a fresh install of magento 2.3 could have these issues.

Maybe someone can help me out?

**Comment by siliconalchemy** [comment]
This is a colossal bug, this issue should not be closed until it's been confirmed fixed in 2.2 and committed to the next point release.

**Comment by magento-engcom-team** [comment]
Hi @N1ghtfl0w. Thank you for your report. 
The issue has been fixed in magento/magento2#19551 by @suneet64 in 2.2-develop branch
Related commit(s):
 - [f0a92f2a2ad7e6a8f6c03b7bded7b2092561f07b](https://github.com/magento/magento2/commit/f0a92f2a2ad7e6a8f6c03b7bded7b2092561f07b)
 - [dbf5293d42f7f88eae47913b746278cf16045914](https://github.com/magento/magento2/commit/dbf5293d42f7f88eae47913b746278cf16045914)
 - [4567a1a95fdd8a4089c808cfd12895de4db1d0e5](https://github.com/magento/magento2/commit/4567a1a95fdd8a4089c808cfd12895de4db1d0e5)

The fix will be available with the upcoming 2.2.9 release.

**Comment by ghost** [comment]
Same issue here, magento 2.3.1

Notice: Undefined index: manufacturer_bucket in /home/myserver/public_html/vendor/magento/module-elasticsearch/SearchAdapter/Aggregation/Builder/Term.php on line 23

Is there a solution for this matter?

**Comment by hostep** [comment]
@koopjesboom: can you please open a new issue with exact steps to reproduce? Because what you are experiencing has nothing to do with this issue, it's best to have one issue focus on one bug, thanks! :)
(your problem might be related to https://github.com/magento/magento2/issues/18802, if that is the case, you can follow what's happening over there)

**Comment by nickpiro** [comment]
We had manufacturer enabled in 2.2.5 and just upgraded to 2.2.8 and we no longer have Manufacturer visible anywhere in product grid or product edit page.

**Comment by gerben86** [comment]
> We had manufacturer enabled in 2.2.5 and just upgraded to 2.2.8 and we no longer have Manufacturer visible anywhere in product grid or product edit page.

Same issue overhere, also @ajusev fix (SQL update) doesn't work at this moment...
